services/web/server/tests/unit/isolated/test_projects__db_utils.py

```diff
@@ -0,0 +1,438 @@
+            access_rights[p] = (
+                p in permissions if invert is False else p not in permissions
+            )
+        return access_rights
+
+    # primary group has needed access, so this should not raise
+    project = {
+        "access_rights": {
+            str(group_id): _project_access_rights_from_permissions(
+                wanted_permissions, invert=False
+            )
+        }
+    }
+    user_groups = [
+        {"type": GroupType.PRIMARY, "gid": group_id},
+        {"type": GroupType.EVERYONE, "gid": 2},
+    ]
+    check_project_permissions(project, user_id, user_groups, wanted_permissions)
+
+    # primary group does not have access, it should raise
+    project = {
+        "access_rights": {
+            str(group_id): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            )
+        }
+    }
+    with pytest.raises(ProjectInvalidRightsError):
+        check_project_permissions(project, user_id, user_groups, wanted_permissions)
+
+    # if no primary group, we rely on standard groups and the most permissive access are used. so this should not raise
+    project = {
+        "access_rights": {
+            str(group_id): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            ),
+            str(group_id + 1): _project_access_rights_from_permissions(
+                wanted_permissions, invert=False
+            ),
+            str(group_id + 2): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            ),
+        }
+    }
+    user_groups = [
+        {"type": GroupType.PRIMARY, "gid": group_id},
+        {"type": GroupType.EVERYONE, "gid": 2},
+        {"type": GroupType.STANDARD, "gid": group_id + 1},
+        {"type": GroupType.STANDARD, "gid": group_id + 2},
+    ]
+    check_project_permissions(project, user_id, user_groups, wanted_permissions)
+
+    # if both primary and standard do not have rights it should raise
+    project = {
+        "access_rights": {
+            str(group_id): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            ),
+            str(group_id + 1): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            ),
+            str(group_id + 2): _project_access_rights_from_permissions(
+                wanted_permissions, invert=True
+            ),
+        }
+    }
+    user_groups = [
+        {"type": GroupType.PRIMARY, "gid": group_id},
+        {"type": GroupType.EVERYONE, "gid": 2},
+        {"type": GroupType.STANDARD, "gid": group_id + 1},
+        {"type": GroupType.STANDARD, "gid": group_id + 2},
+    ]
+    with pytest.raises(ProjectInvalidRightsError):
+        check_project_permissions(project, user_id, user_groups, wanted_permissions)
+
+    # the everyone group has access so it should not raise
```

Comment:
TYPO: you meant gid? instead of git?

Reply:
? gid as in group id? well anyway this test did not change.